stats/opentelemetry/grpc_trace_bin_propagator_test.go

```diff
@@ -67,23 +67,23 @@ func (s) TestInject(t *testing.T) {
 			defer cancel()
 			ctx = oteltrace.ContextWithSpanContext(ctx, test.injectSC)
 
-			c := itracing.NewCustomCarrier(ctx, itracing.Outgoing)
+			c := itracing.NewIncomingCarrier(ctx)
```

Comment:
You changed this from `Outgoing` to `NewIncomingCarrier` (?!)

Reply:
Ah yeah. Code completion.... :(. But good that you caught it. It didn't fail in propagator test because direction don't matter if we are not fetching keys. I have added the test to verify keys as well. Although in case of GrpcTraceBinPropagator the keys will always be just `grpc-trace-bin`. However, we still should verify the keys from carrier.

Reply:
Hmmm, should we change that?  If someone makes an outgoing carrier and then calls Extract, should that work by reading the outgoing metadata?  I'd assume so.

Reply:
Do you mean fail if someone calls Extract after creating an outgoing carrier? If someone calls extract with outgoing carrier, they will not get the trace context because we `ValueFromIncomingContext` for Get. Perhaps we can document that carrier with incoming context works.

Reply:
or better we should just let the constructor create the incoming context and outgoing based on the carrier called. Like if you call `NewIncomingCarrier`, we can just do
```
md, ok := metadata.FromIncomingContext(ctx)
	if !ok {
		return &Carrier{ctx: metadata.NewIncomingContext(ctx, metadata.MD{})
	}
	return &Carrier{ctx: metadata.NewIncomingContext(ctx, md)}
```

That way we don't need direction at all and it will keep things simple and restricted.

Reply:
I have updated the constructor and documentation. Now its still possible for someone to call `Extract` with outgoing carrier or vice versa but we have documentation to call out that  its not the correct way. However, they are not allowed to create outgoing carrier with incoming metadata or vice versa. Let me know if that looks good.

Reply:
Actually, scratch that. I reverted it becuase it compromises the efficiency due to copying back and forth. For now, i have kept the documentation. I don't anything can stop from calling extract with outgoing carrier. If an outgoing carrier is initialized with context that has `grpc-trace-bin` header, it will just work fine except that the `Keys()` will return outgoing keys. 

I have added the documentation in carrier though that Incoming carrier has to be used with Extract and outgoing with Inject. Also, its internal to grpc so may be fine?

Reply:
As discussed offiline, I added separate incoming and outgoing carrier and they only implement Get() and Set() respectively and log error for Set() and Get() respectively.